### PR TITLE
Enable performance measurement in INT8 calibration unit test

### PR DESCRIPTION
**Pull request by hshen14**
This PR is to support performance measurement in INT8 calibration test. After unit test, it will print the throughput (images/second), latency (milliseconds), and accuracy.

test=develop

### CODE REVIEW

File: python/paddle/fluid/contrib/tests/test_calibration.py [hunk 1]
```diff
@@ -225,15 +231,25 @@ def run_program(self, model_path, generate_int8=False, algo='direct'):
...
                 format(os.path.abspath("calibration_out")))
         else:
-            return np.sum(test_info) / cnt
+            throughput = cnt / np.sum(periods)
+            latency = np.average(periods)
+            acc1 = np.sum(test_info) / cnt
+            return (throughput, latency, acc1)
 
     def test_calibration(self):
         self.download_resnet50_model()
```
Comment (luotao1) [inline python/paddle/fluid/contrib/tests/test_calibration.py]: `self.download_resnet50_model()`-》`self.download_model()`

File: python/paddle/fluid/contrib/tests/test_calibration.py [hunk 2]
```diff
@@ -246,11 +262,18 @@ def download_mobilenetv1_model(self):
...
-        int8_acc1 = self.run_program("calibration_out")
+        (int8_throughput, int8_latency,
+         int8_acc1) = self.run_program("calibration_out")
         delta_value = np.abs(fp32_acc1 - int8_acc1)
         self.assertLess(delta_value, 0.01)
+        print("FP32 MobileNet-V1: throughput {0}, latency {1}, accuracy {2}".
+              format(fp32_throughput, fp32_latency, fp32_acc1))
+        print("INT8 MobileNet-V1: throughput {0}, latency {1}, accuracy {2}".
+              format(int8_throughput, int8_latency, int8_acc1))
+        sys.stdout.flush()
```
Comment (luotao1) [inline python/paddle/fluid/contrib/tests/test_calibration.py]: line 263-276 could be removed after re-define `self.download_model()`
```
class TestCalibrationForMobilenetv1(TestCalibrationForResnet50):
    def download_model(self):
            ...
```

Comment (luotao1) [reply python/paddle/fluid/contrib/tests/test_calibration.py]: Besides, do you need an `args` to test the full dataset in local machine?
```
def download_model(self):
   if args.dataset='full':
        xxx
   else
        xxx
```
Then, you can write in the README
```
python python/paddle/fluid/contrib/tests/test_calibration.py --dataset=full
```

File: python/paddle/fluid/contrib/tests/test_calibration.py (hunk 1 above)
Comment (hshen14) [reply python/paddle/fluid/contrib/tests/test_calibration.py]: Done

File: python/paddle/fluid/contrib/tests/test_calibration.py (hunk 2 above)
Comment (hshen14) [reply python/paddle/fluid/contrib/tests/test_calibration.py]: Done for code reuse. On args, I did not find a good way to pass args in unit test, so I used python environment to control the dataset:  `DATASET=small/full python python/paddle/fluid/contrib/tests/test_calibration.py`.

**Comment by luotao1** [comment]
> DATASET=small/full

It's wrong in use.
```
$ ctest -R test_calibration -V
UpdateCTestConfiguration  from :/home/luotao/Paddle/cpu_build/DartConfiguration.tcl
UpdateCTestConfiguration  from :/home/luotao/Paddle/cpu_build/DartConfiguration.tcl
Test project /home/luotao/Paddle/cpu_build
Constructing a list of tests
Done constructing a list of tests
Checking test dependency graph...
Checking test dependency graph end
test 480
    Start 480: test_calibration

480: Test command: /usr/local/bin/cmake "-E" "env" "FLAGS_init_allocated_mem=true" "FLAGS_cudnn_deterministic=true" "FLAGS_cpu_deterministic=true" "FLAGS_limit_of_tmp_allocation=4294967296" "PYTHONPATH=/home/luotao/Paddle/cpu_build/python" "/home/luotao/.jumbo/bin/python2.7" "-u" "test_calibration.py"
480: Test timeout computed to be: 600
480: Cache file /home/luotao/.cache/paddle/dataset/int8/download/calibration_test_data.tar.gz not found, downloading http://paddle-inferenc[==================================================].gz
480: E
480: E
480: ======================================================================
480: ERROR: test_calibration (__main__.TestCalibrationForMobilenetv1)
480: ----------------------------------------------------------------------
480: Traceback (most recent call last):
480:   File "test_calibration.py", line 135, in setUp
480:     self.infer_iterations = 50000 if os.environ["DATASET"] == 'full' else 50
480:   File "/home/luotao/.jumbo/lib/python2.7/UserDict.py", line 23, in __getitem__
480:     raise KeyError(key)
480: KeyError: 'DATASET'
480:
480: ======================================================================
480: ERROR: test_calibration (__main__.TestCalibrationForResnet50)
```

**Comment by luotao1** [comment]
please update setup.py.in in another PR to package the int8 into whl.
https://github.com/PaddlePaddle/Paddle/blob/7bc8481c62bc0d88e495689894fb481ae798996b/python/setup.py.in#L109-L118
when I run 
```
make install -j
pip install xxx.whl
python python/paddle/fluid/contrib/tests/test_calibration.py
```
The result is:
```
Traceback (most recent call last):
  File "python/paddle/fluid/contrib/tests/test_calibration.py", line 28, in <module>
    import int8_inference.utility as int8_utility
ImportError: No module named int8_inference.utility
```

File: python/paddle/fluid/contrib/tests/test_calibration.py [hunk 3]
```diff
@@ -19,10 +19,8 @@
 import random
 import paddle
 import paddle.fluid as fluid
-import argparse
 import functools
 import contextlib
-import paddle.fluid.profiler as profiler
 from paddle.dataset.common import download
 from PIL import Image, ImageEnhance
```
Comment (luotao1) [inline python/paddle/fluid/contrib/tests/test_calibration.py:25]: please update line 44 '# TODO(guomingz): Remove duplicated code from line 45 ~ line 114'->
'# TODO(guomingz): Remove duplicated code from resize_short, crop_image, process_image, _reader_creator'.
The reason is that the line number changes in this PR.

Comment (hshen14) [reply python/paddle/fluid/contrib/tests/test_calibration.py:25]: Done, thanks!

**Comment by hshen14** [comment]
> please update setup.py.in in another PR to package the int8 into whl.
> [Paddle/python/setup.py.in](https://github.com/PaddlePaddle/Paddle/blob/7bc8481c62bc0d88e495689894fb481ae798996b/python/setup.py.in#L109-L118)
> 
> Lines 109 to 118 in [7bc8481](/PaddlePaddle/Paddle/commit/7bc8481c62bc0d88e495689894fb481ae798996b)
> 
>  'paddle.fluid.contrib', 
>  'paddle.fluid.contrib.decoder', 
>  'paddle.fluid.contrib.quantize', 
>  'paddle.fluid.contrib.reader', 
>  'paddle.fluid.contrib.slim', 
>  'paddle.fluid.contrib.slim.core', 
>  'paddle.fluid.contrib.slim.graph', 
>  'paddle.fluid.contrib.slim.prune', 
>  'paddle.fluid.contrib.slim.quantization', 
>  'paddle.fluid.contrib.utils', 
> 
> when I run
> ```
> make install -j
> pip install xxx.whl
> python python/paddle/fluid/contrib/tests/test_calibration.py
> ```
> The result is:
> 
> ```
> Traceback (most recent call last):
>   File "python/paddle/fluid/contrib/tests/test_calibration.py", line 28, in <module>
>     import int8_inference.utility as int8_utility
> ImportError: No module named int8_inference.utility
> ```

**Comment by luotao1** [comment]
After https://github.com/PaddlePaddle/Paddle/pull/15560/commits/e4beac4519599d8c6052d6949593778d52d24579, I test it on a old machine
```
    Start 480: test_calibration

480: Test command: /usr/local/bin/cmake "-E" "env" "FLAGS_init_allocated_mem=true" "FLAGS_cudnn_deterministic=true" "FLAGS_cpu_deterministic=true" "FLAGS_limit_of_tmp_allocation=4294967296" "PYTHONPATH=/home/luotao/Paddle/cpu_build/python" "/home/luotao/.jumbo/bin/python2.7" "-u" "test_calibration.py"
480: Test timeout computed to be: 600
480:
480: Cache file /home/luotao/.cache/paddle/dataset/int8/download/mobilenetv1_int8_model.tar.gz not found, downloading http://paddle-inferen[==================================================]ar.gz
480: Start calibration ...
480: Calibration is done and the corresponding files are generated at /home/luotao/Paddle/cpu_build/python/paddle/fluid/contrib/tests/calibration_out
480: FP32 MobileNet-V1: throughput 12.4694766719, latency 0.0801958274841, accuracy 0.62
480: INT8 MobileNet-V1: throughput 1.57680277844, latency 0.634194722176, accuracy 0.62
480: .
480: Cache file /home/luotao/.cache/paddle/dataset/int8/download/resnet50_int8_model.tar.gz not found, downloading http://paddle-inference-[==================================================]
480: Start calibration ...
480: Calibration is done and the corresponding files are generated at /home/luotao/Paddle/cpu_build/python/paddle/fluid/contrib/tests/calibration_out
480: FP32 ResNet-50: throughput 2.3786103997, latency 0.420413532257, accuracy 0.72
480: INT8 ResNet-50: throughput 0.787770993066, latency 1.26940444469, accuracy 0.72
480: .
480: ----------------------------------------------------------------------
480: Ran 2 tests in 287.085s
480:
480: OK
```
There are some problem:
- The accuracy is not the same in https://github.com/PaddlePaddle/Paddle/pull/15548/files. Is the accuracy different among machines?
- Could you add more print log, such as
```
Start calibration ResNet-50 ...
Using 50 images. / Using 50000 images.
FP32 ResNet-50: throughput 2.3786103997 images/second, latency 0.420413532257 second, accuracy 0.72
```

**Comment by hshen14** [comment]
> After [e4beac4](https://github.com/PaddlePaddle/Paddle/commit/e4beac4519599d8c6052d6949593778d52d24579), I test it on a old machine
> 
> ```
>     Start 480: test_calibration
> 
> 480: Test command: /usr/local/bin/cmake "-E" "env" "FLAGS_init_allocated_mem=true" "FLAGS_cudnn_deterministic=true" "FLAGS_cpu_deterministic=true" "FLAGS_limit_of_tmp_allocation=4294967296" "PYTHONPATH=/home/luotao/Paddle/cpu_build/python" "/home/luotao/.jumbo/bin/python2.7" "-u" "test_calibration.py"
> 480: Test timeout computed to be: 600
> 480:
> 480: Cache file /home/luotao/.cache/paddle/dataset/int8/download/mobilenetv1_int8_model.tar.gz not found, downloading http://paddle-inferen[==================================================]ar.gz
> 480: Start calibration ...
> 480: Calibration is done and the corresponding files are generated at /home/luotao/Paddle/cpu_build/python/paddle/fluid/contrib/tests/calibration_out
> 480: FP32 MobileNet-V1: throughput 12.4694766719, latency 0.0801958274841, accuracy 0.62
> 480: INT8 MobileNet-V1: throughput 1.57680277844, latency 0.634194722176, accuracy 0.62
> 480: .
> 480: Cache file /home/luotao/.cache/paddle/dataset/int8/download/resnet50_int8_model.tar.gz not found, downloading http://paddle-inference-[==================================================]
> 480: Start calibration ...
> 480: Calibration is done and the corresponding files are generated at /home/luotao/Paddle/cpu_build/python/paddle/fluid/contrib/tests/calibration_out
> 480: FP32 ResNet-50: throughput 2.3786103997, latency 0.420413532257, accuracy 0.72
> 480: INT8 ResNet-50: throughput 0.787770993066, latency 1.26940444469, accuracy 0.72
> 480: .
> 480: ----------------------------------------------------------------------
> 480: Ran 2 tests in 287.085s
> 480:
> 480: OK
> ```
> There is some problem:
> 
> * The accuracy is not the same in https://github.com/PaddlePaddle/Paddle/pull/15548/files. Is the accuracy different among machines?
> * Could you add more print log, such as
> 
> ```
> Start calibration ResNet-50 ...
> Using 50 images. / Using 50000 images.
> FP32 ResNet-50: throughput 2.3786103997 images/second, latency 0.420413532257 second, accuracy 0.72
> ```

Sure, I will add more print log. 0.62 and 0.72 is expected. What do you mean the difference?

**Comment by luotao1** [comment]
> What do you mean the difference

Sorry, I read the result upside down

> add more print log

For full dataset, since it costs nearly one hour to run the FP32 acc. Could you print some log for this, such as 
```
100 images,
200 images,
...
```

**Comment by hshen14** [comment]
> > What do you mean the difference
> 
> Sorry, I read the result upside down
> 
> > add more print log
> 
> For full dataset, since it costs nearly one hour to run the FP32 acc. Could you print some log for this, such as
> 
> ```
> 100 images,
> 200 images,
> ...
> ```

Done

**Comment by hshen14** [comment]
@luotao1 

[09:46:00] :	 [Step 1/1] The following tests FAILED:
[09:46:00] :	 [Step 1/1] 	197 - test_eager_deletion_transformer (Failed)
[09:46:00]W:	 [Step 1/1] Errors while running CTest

http://ci.paddlepaddle.org/viewLog.html?buildId=54668&buildTypeId=PaddleMac_MacPrCi&tab=buildLog

**Comment by luotao1** [comment]
please add log for the location of download data and model.
```
Cache file /home/luotao/.cache/paddle/dataset/int8/download/mobilenetv1_int8_model.tar.gz not found, downloading http://paddle-inferen[==================================================]ar.gz
```
If we rerun again, we will not know where the location of data and model .

**Comment by luotao1** [comment]
> test_eager_deletion_transformer (Failed)

Please rerun again, I have already reported this randon fail.

**Comment by hshen14** [comment]
> please add log for the location of download data and model.
> 
> ```
> Cache file /home/luotao/.cache/paddle/dataset/int8/download/mobilenetv1_int8_model.tar.gz not found, downloading http://paddle-inferen[==================================================]ar.gz
> ```
> If we rerun again, we will not know where the location of data and model .

Done

**Comment by luotao1** [comment]
```
[04:28:58][Step 2/5] CUSTOMBUILD : CMake error : Cannot restore timestamp D:/home/BuildAgent/work/a9b0372f0aea0a80/build/paddle/fluid/inference/tests/book/CMakeFiles/generate.stamp [D:\home\BuildAgent\work\a9b0372f0aea0a80\build\paddle\fluid\inference\tests\book\test_inference_nlp.vcxproj]
```
http://ci.paddlepaddle.org/viewLog.html?buildId=55056&tab=buildLog&buildTypeId=PaddleWindows_PrWindowsCi&logTab=tree&filter=all&_focus=2548

**Review by luotao1 — APPROVED** [review]

**Pull request merged**